### PR TITLE
Infighting update: triggers and monster factions

### PR DESCRIPTION
**Pull request by Coolthulhu**
Allows monsters to get mad at non-player targets. Moves PLAYER_NEAR monster trigger processing to monster::plan and renames it to MTRIG_HOSTILE_NEAR.

Added 2 new behaviors: swarming and group morale.
Swarming causes idle (target-less) monsters to go towards nearest ally or move away from the nearest ally that's too near.
Group morale boosts creature's morale when near allies.
Added both behaviors to wasps and bees and lowered their morale to make group morale effect show.

Monster factions are first assigned to monster types: when loading an mtype, its json will be checked for "default_faction" member. If not found, first species' name will be used as faction's name.

Generated monster acquires its faction from its type, but the faction is not tied to mtype after that.
Monsters keep their factions as ids and translate them to/from strings when saving/loading.

In-game effects will be:
- Zed scientists will stop getting killed by their own manhacks (will spam them hard - maybe they should get them as ammo?)
- Violent animals will fight back when attacked by zombies
- Bees and wasps will prefer to fight in groups
- Fixed a bug that prevented animals from fleeing from hostile monsters when the player or an NPC is visible (didn't catch that at first because I used Debug Invisibility when testing)

Json stuff that could be done with it:
- The possibility of splitting species into separate factions, like wolves and rabbits
- Multi-species factions - blobs not fighting with zeds
- Adding stuff to player's faction (not exactly the same effect as friendly = -1)

Possible further development:
- Monster faction relations (currently it's all "us vs them")
- Multi-faction monsters (police zeds allied with zeds and policebots)
- (Much later on) merging with NPC factions

### CODE REVIEW

**Comment by BevapDin** [comment]
Why don't just s store a pointer to a const `mon_faction` object instead of a `faction_id`, like this:
```
struct mon_faction {
    int id;
    std::string id_as_string;
};
```
Than compare the pointer itself, that is as fast as comparing the int. That gets you rid of the inverted map and you can even add more properties to that faction.

**Comment by Coolthulhu** [comment]
Removing [WIP] because it works already.

I noticed that many monsters don't have the sense of sight. Unfortunately the entire infighting system depends on sight.
I changed the minimal sight range to 1. This doesn't change mechanics of monsters attacking the player much (adjacent monsters would usually attack already), but enables many types of monsters (ants) to fight back when attacked by zombies.

I gave bees and wasps range 10 sight to show the new swarming behavior. Could also work nicely on spiders.

File: src/event.cpp [hunk 1]
```diff
@@ -293,7 +291,6 @@ void event::per_turn()
    // About once every 10 minutes. Suppress in classic zombie mode.
    if (g->levz >= 0 && one_in(100) && !ACTIVE_WORLD_OPTIONS["CLASSIC_ZOMBIES"]) {
     monster eyebot(GetMType("mon_eyebot"));
-    eyebot.faction_id = faction_id;
```
Comment (KA101) [inline src/event.cpp]: Should probably only have this summon one eyebot: cut down on the swarming folks report.

File: src/monstergenerator.cpp [hunk 2]
```diff
@@ -237,8 +239,8 @@ void MonsterGenerator::init_trigger()
     trigger_map["NULL"] = MTRIG_NULL;// = 0,
     trigger_map["STALK"] = MTRIG_STALK;//  // Increases when following the player
     trigger_map["MEAT"] = MTRIG_MEAT;//  // Meat or a corpse nearby
-    trigger_map["PLAYER_WEAK"] = MTRIG_PLAYER_WEAK;// // The player is hurt
-    trigger_map["PLAYER_CLOSE"] = MTRIG_PLAYER_CLOSE;// // The player gets within a few tiles
+    trigger_map["PLAYER_WEAK"] = MTRIG_HOSTILE_WEAK;// // Hurt hostile player/npc/monster seen
```
Comment (KA101) [inline src/monstergenerator.cpp]: Still a TODO.

**Comment by Rivet-the-Zombie** [comment]
I've always wondered why our giant insects were all blind. It never made any sense to me.

**Comment by KA101** [comment]
Was supposed to reference their reliance on smell, IIRC.  No matter.

**Comment by Coolthulhu** [comment]
I don't know how to end an event.
I think police bots deserve their own separate PR for eyebot balancing, but also for proper copbot spawning (they always spawn at exact same spot relative to player now).

The whole player_weak changed to hostile_weak is not in yet, yeah. I'll write it in a moment, brb

**Comment by Coolthulhu** [comment]
OK, it's in, but it has the same weakness as the player implementation - it will often cause critters to switch to "tracking" when in danger (like animals hurt by player often do) rather than fleeing.
This is more of a problem with creatures balancing anger vs morale and isn't directly related to infighting, but is noticeable here.

MTRIG_HOSTILE_WEAK should function in more or less the same way MTRIG_PLAYER_WEAK does when the player is the target.

**Comment by KA101** [comment]
>it will often cause critters to switch to "tracking" when in danger (like animals hurt by player often do) rather than fleeing.

OK, so next PR should have Fleeing automatic at (say) 30% of HP?

**Comment by Coolthulhu** [comment]
30% hp sounds OK.
Or maybe a MTRIG_ME_WEAK working like a reverse HOSTILE_WEAK trigger and making hurt critters accumulate negative anger (though this could result in perma-scared wolves).

**Comment by KA101** [comment]
Yeah, we should probably implement critters healing over time at some point.

**Comment by KA101** [comment]
One thing I noticed: bees are gonna fight zombies right off thanks to the scarred zeds (which, incidentally, they kill without significant damage--possible rebalance/gear tweaks needed later).  Not sure if the resulting Hostile screws up the swarming.

**Comment by Coolthulhu** [comment]
Swarming only happens when there are no targets nearby, but bees are nearly blind (10 tiles, was 0 before), so they should regroup after they kill stuff.

Making bees not kill scared zeds will probably require faction relations.

**Comment by KA101** [comment]
>Making bees not kill scared zeds will probably require faction relations.

More like they've stung the thing a LOT and it just hasn't died, I thought. They aren't actually friendly, just suspension-of-disbelief that the scars built over (presumably) five+ days of trying to sting it to death are surmounted in about three minutes once the player arrives.

But this does make bees pretty effective.

File: src/monmove.cpp [hunk 3]
```diff
@@ -144,70 +166,105 @@ void monster::plan(const mfactions &factions)
...
-                            wandy = posy() * 2 - mon.posy();
-                            wandf = 40;
-                        } else {
-                            closest = -3 - i;
-                            selected_slope = bresenham_slope;
-                        }
-                    }
+    // Friendly monsters here
+    // Avoid for hordes of same-faction stuff or it could get expensive
+    auto myfaction = factions.find( faction->id )->second;
```
Comment (BevapDin) [inline src/monmove.cpp]: This is inconsistent with the changes in game.cpp. If the monster is friendly, there will be an entry -1 in the factions map, but no entry with `faction->id` of the monster itself. The find will return an invalid iterator, which is dereferenced and will cause a crash.

Might be related to #10895 as it crashes at this line.

Comment (KA101) [inline src/monmove.cpp]: Lovely.

**Pull request merged**